main.go

```diff
@@ -1505,3 +1551,121 @@ func main() {
 		os.Exit(1)
 	}
 }
+
+// testOutputBuf is used to buffer the output of concurrent tests.
+type testOutputBuf struct {
+	mu             sync.Mutex
+	output         []outputEntry
+	stdout, stderr io.Writer
+	outerr, errerr error
+	done           chan struct{}
+}
+
+// flush the output to stdout and stderr.
+// This waits until done is closed.
+func (b *testOutputBuf) flush(stdout, stderr io.Writer) error {
+	b.mu.Lock()
+
+	var err error
+	b.stdout = stdout
+	b.stderr = stderr
+	for _, e := range b.output {
+		var w io.Writer
+		var errDst *error
+		if e.stderr {
+			w = stderr
+			errDst = &b.errerr
+		} else {
+			w = stdout
+			errDst = &b.outerr
+		}
+		if *errDst != nil {
+			continue
+		}
+
+		_, werr := w.Write(e.data)
+		if werr != nil {
+			if err == nil {
+				err = werr
+			}
+			*errDst = err
+		}
+	}
+
+	b.mu.Unlock()
+
+	<-b.done
+
+	return err
+}
+
+// testStdout writes stdout from a test to the output buffer.
+type testStdout testOutputBuf
+
+func (out *testStdout) Write(data []byte) (int, error) {
+	buf := (*testOutputBuf)(out)
+	buf.mu.Lock()
+
+	if buf.stdout != nil {
+		// Write the output directly.
+		err := out.outerr
+		buf.mu.Unlock()
+		if err != nil {
+			return 0, err
+		}
+		return buf.stdout.Write(data)
+	}
+
+	defer buf.mu.Unlock()
+
+	// Append the output.
+	var prev []byte
+	if len(buf.output) > 0 && !buf.output[len(buf.output)-1].stderr {
+		prev = buf.output[len(buf.output)-1].data
+		buf.output = buf.output[:len(buf.output)-1]
+	}
+	buf.output = append(buf.output, outputEntry{
+		stderr: false,
+		data:   append(prev, data...),
+	})
```

Comment:
I think this can be simplified to the following:

```suggestion
if len(buf.output) == 0 || !buf.output[len(buf.output)-1].stderr {
    buf.output = append(buf.output, outputEntry{
        stderr: false,
    })
}
last := buf.output[len(buf.output)-1]
last.data = append(last.data, data...)
```

At least I think it's simpler :shrug: essentially it first makes sure the last output entry is of the current output type (stdout, stderr) and then just appends to that output entry.

Reply:
There are a few typos here, but I did manage to implement a version of this tweak